### DISCUSSION TITLE
Expo Notification with Image?

### DISCUSSION THREAD

**Discussion by NilsBaumgartner1994** [expo/expo#27980] [Archived: Questions]
How to send an image in the expo push notification message?

```
const messages = expoPushTokens.map(token => ({
            to: token,
            sound: 'default',
            badge: payload?.ios_badge_count || 0,
            title: title, // Replace with the actual field name
            body: body, // Replace with the actual field name
            data: data, // Replace with the actual field name
            /**
             attachments: {
             identifier: string | null;
             url: string | null;
             type: string | null;
             }[];
             */
            attachments: [
                {
                    identifier: 'image',
                    url: 'https://www.fnordware.com/superpng/pnggrad8rgb.png',
                    type: 'image/png'
                }
            ]
        }));

        //console.log("Messages:")
        //console.log(messages)

        try {
            await axios.post('https://exp.host/--/api/v2/push/send', messages);
        } catch (e) {
            console.log(`Failed to send notification: ${e.message}`)
            input.status = 'failed';
            input.status_log = e.message.toString();
            throw new Error(`Failed to send notification: ${e.message}`);
        }
```

**Comment by williambroberts** [comment]
Does this work for you? The image doesn't display on the iPhone for me.

**Reply by NilsBaumgartner1994** [reply]
Well, that’s exactly what my issue is.

**Reply by williambroberts** [reply]
when I  console.warn(notification.request.content); i see attachments:[], empty array. Is it the same for you?

**Reply by NilsBaumgartner1994** [reply]
Haven’t tested that yet. But I think expo currently just can’t send images with push notifications

**Reply by thunder6230** [reply]
as far as i know expo notification doesn't support image (no idea why)

**Reply by jyaconelli** [reply]
Also having this issue. Seems like this should be a basic feature of Expo

**Reply by sregg** [reply]
@NilsBaumgartner1994 did you figure out how to include an image in the notification?

**Reply by NilsBaumgartner1994** [reply]
> @NilsBaumgartner1994 did you figure out how to include an image in the notification?

Sadly no.

**Comment by NilsBaumgartner1994** [comment]
As I understood correctly, it is not possible with expo. You would need to
send them via fcm directly or Apple.

Simon Reggiani ***@***.***> schrieb am Fr. 19. Juli 2024 um
15:52:

**Comment by BryanMwangi** [comment]
Anyone have any idea if image support will be included in future?

**Comment by BrodaNoel** [comment]
Is New Arch going to help to achieve this? Maybe plugins? something? Or is it something that is never going to be implemented ever?

**Comment by joekendal** [comment]
+1

**Comment by kourtc** [comment]
+1

**Comment by abdu-zeyad** [comment]
+1

**Comment by a-eid** [comment]
wondering if there are any updates on this ?

**Comment by sergiopinhon** [comment]
I think it would be nice! +1 too!!!

**Comment by amorimcode** [comment]
+1 this is needed

**Comment by HiltonLuu** [comment]
+1

**Comment by CavalcanteLeo** [comment]
I heard somewhere that if you follow this guide https://rnfirebase.io/messaging/ios-notification-images, the API will be able to push notifications with images. I haven't tested it, but it might be worth a try.

**Reply by BrodaNoel** [reply]
As you can see there, it may imply using a plugin (in case it was not created yet). So, it wouldn't be a plug-and-play library.

Let us know if you had some luck with it

**Reply by MasterBhuvnesh** [reply]
let me check it out

**Comment by UsamaAshraf82** [comment]
+1

**Comment by tn531** [comment]
+1

**Comment by abisarwan-go** [comment]
+1

**Comment by ahmed1hisham** [comment]
+1

**Comment by dohooo** [comment]
same

**Comment by golbin** [comment]
🆙

**Comment by golbin** [comment]
And, it would be nice to have avatar image support.

**Comment by FlawaCLV** [comment]
+1

I understand expo wants to prevent local/embedded notifications to save data & money, but at least remote image using an URL seems pretty straight forward to implement.

Anyone tried [Notifee](https://notifee.app)?

**Reply by UsamaAshraf82** [reply]
Yes tried it but cannot make it work on iOS

**Comment by lostb1t** [comment]
+1

**Comment by herrrta** [comment]
+1

**Comment by ftr9** [comment]
+1

**Comment by zlatiz** [comment]
+1

**Comment by medivmed** [comment]
+1 any news on this topic in expo api doc they are saying that we could send image via api using richContent object... but tested it and it is not working for me
![image](https://github.com/user-attachments/assets/51afd40d-9bab-461f-bb53-9363d6ea4f15)

**Reply by daniel-soerensen-cg** [reply]
Thank you so much for posting this, @medivmed!
I just tried it, too, and here it is working.

My test payload looks like this: 

`{
  "to": "ExponentPushToken[...]",
  "richContent": { "image": "http://192.168.4.242:8083/assets/logo.png" },
  "title": "Nachricht",
  "body": "Hello There... !",
}
`

And the image is displayed: 
![image](https://github.com/user-attachments/assets/c695e254-df6d-4972-b31d-dd0a58c3ce13)

**Reply by NilsBaumgartner1994** [reply]
> Thank you so much for posting this, @medivmed! I just tried it, too, and here it is working.
> 
> My test payload looks like this:
> 
> `{ "to": "ExponentPushToken[...]", "richContent": { "image": "http://192.168.4.242:8083/assets/logo.png" }, "title": "Nachricht", "body": "Hello There... !", } `
> 
> And the image is displayed: ![image](https://private-user-images.githubusercontent.com/194486865/437018604-c695e254-df6d-4972-b31d-dd0a58c3ce13.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDU1MDk3NzYsIm5iZiI6MTc0NTUwOTQ3NiwicGF0aCI6Ii8xOTQ0ODY4NjUvNDM3MDE4NjA0LWM2OTVlMjU0LWRmNmQtNDk3Mi1iMzFkLWRkMGE1OGMzY2UxMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQyNFQxNTQ0MzZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05NWZiMDI1MjUzZTNlZDJmN2MwNmQ5ZGU1MjYyZmY2ZWI3NzcxNTY4MmJjNzhmMmI4NDYxZWY1NzhiY2ZhNmQ2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.N-PW-ep5ucZj-x4qFzbQy-OOBl_FuhvWTq6-3GzBIrM)

Nice solution for android. But iOS seems not to support that?

**Reply by daniel-soerensen-cg** [reply]
Yeah, as far as I know, the Android OS supports displaying images in notifications out of the box (so imho Expo just needed to pass this data through), and iOS does not. 
I looked for a solution for iOS a while ago (as I was still using APNS servers directly) and I found a way to show images in iOS, too, but some native code was needed to make it work, so the overhead was too much for me.

BTW: Thanks Expo team for adding this :)

**Comment by lukhol** [comment]
+1

**Comment by hammamikhairi** [comment]
If the JSON body is a single message object, the server won't process `richContent` key (and some others) for some reason :/

to fix this, simply send the payload as an array of message objects, even if it's a single one and that should fix it.
example : 

```python
def send_push_with_image(expo_push_token, body, image_url):
    message = [
        {
            "to": expo_push_token,
            "body": body,
            "richContent": {
                "image": image_url
            }
        }
    ]

    response = requests.post(
        "https://exp.host/--/api/v2/push/send",
        headers={
            "accept": "application/json",
            "content-type": "application/json",
        },
        data=json.dumps(message)
    )

    if response.status_code not in (200, 201):
        raise Exception(f"Push failed: {response.status_code} {response.text}")

    print("Push with image sent!")
```

++ image size can not surpass 1mb

**Comment by JClackett** [comment]
Any updates? surely this is a pretty important thing to make an expo app fully featured - how many apps use custom notifications images

**Reply by MasterBhuvnesh** [reply]
yes you can refer this 

```
def send_push_with_image(expo_push_token, body, image_url):
    message = [
        {
            "to": expo_push_token,
            "body": body,
            "richContent": {
                "image": image_url
            }
        }
    ]

    response = requests.post(
        "https://exp.host/--/api/v2/push/send",
        headers={
            "accept": "application/json",
            "content-type": "application/json",
        },
        data=json.dumps(message)
    )

    if response.status_code not in (200, 201):
        raise Exception(f"Push failed: {response.status_code} {response.text}")

    print("Push with image sent!")
```

**Reply by MasterBhuvnesh** [reply]
Right now the app that can use expo notification with images .The app which are build using expo 53

**Comment by NilsBaumgartner1994** [comment]
But also for iOS with Image?

Bhuvnesh Verma ***@***.***> schrieb am So. 27. Juli 2025 um
11:54:

**Reply by MasterBhuvnesh** [reply]
i haven't tried it but i guess we need to setup some kind of Notification Service Extension target

**Reply by Navipro70** [reply]
It won't help for expo-notifications, tested with firebase messaging, remote notifications are showing images but locally scheduled notifications with expo-notifications doesn't

**Comment by Pushpajekumar** [comment]
const messages: NotificationPayload[] = validTokens.map(
      (token: string) => ({
        to: token,
        sound: sound || "default",
        title,
        body: content,
        richContent: {
          image: image || "",
        },
      })
    );



This works fine

**Comment by DevGeekPhoenix** [comment]
For anyone still hitting this: `richContent.image` works on Android today but only renders as a small thumbnail (`setLargeIcon`), not a large banner. I've opened a proposal to make it render as a full expanded image (`BigPictureStyle`), plus iOS attachment support:

https://github.com/expo/expo/discussions/46973

If you want this, a 👍 or comment there would help it get on the maintainers' radar.

**Comment by NilsBaumgartner1994** [comment]
Any Updates on this? What Are the Problems with that?
